`partial_apply` support for coroutines

asavonic opened the pull request:
The patch adds lowering of partial_apply instructions for coroutines. 
This pattern seems to trigger a lot of type mismatch errors in IRGen, because coroutine functions are not substituted in the same way as regular functions (see the patch https://github.com/apple/swift/commit/07f03bd287b5f5297ef7a30a49622d0ad29243eb "Use pattern substitutions to consistently abstract yields" for more details). The odd type conversions in the patch are related to this issue, and these should be checked carefully. Perhaps it is better to enable substitutions for coroutine functions instead (at least for some cases).

Other than that, lowering of partial_apply for coroutines is straightforward: we generate another coroutine that captures arguments passed to the partial_apply instructions. It calls the original coroutine for yields (first return) and yields the resulting values. Then it calls the original function's continuation for return or unwind, and forwards them to the caller as well.

After IRGen, LLVM's Coroutine pass transforms the generated coroutine (along with all other coroutines) and eliminates llvm.coro.* intrinsics. LIT tests check LLVM IR after this transformation.

Test cases are taken from partial_apply.sil test, and adapted to use coroutines.

asavonic committed:
Partial apply for coroutines

The patch adds lowering of partial_apply instructions for coroutines.

This pattern seems to trigger a lot of type mismatch errors in IRGen, because
coroutine functions are not substituted in the same way as regular functions
(see the patch 07f03bd2 "Use pattern substitutions to consistently abstract
yields" for more details). The odd type conversions in the patch are related to
this issue, and these should be checked carefully. Perhaps it is better to
enable substitutions for coroutine functions instead (at least for some cases).

Other than that, lowering of partial_apply for coroutines is straightforward: we
generate another coroutine that captures arguments passed to the partial_apply
instructions. It calls the original coroutine for yields (first return) and
yields the resulting values. Then it calls the original function's continuation
for return or unwind, and forwards them to the caller as well.

After IRGen, LLVM's Coroutine pass transforms the generated coroutine (along with
all other coroutines) and eliminates llvm.coro.* intrinsics. LIT tests check
LLVM IR after this transformation.

asl commented:
@rjmccall Will you please take a look on https://github.com/apple/swift/pull/71653/commits/ee57131dc465eb0e7fec5c5c5b7f12d90d29c7e8 ?

asl commented:
Please test with https://github.com/apple/llvm-project/pull/8199
@swift-ci please test

asavonic committed:
Handle begin_apply in PartialApplyCombiner

asl commented:
@swift-ci please test

asl commented:
Please test with https://github.com/apple/llvm-project/pull/8199
@swift-ci please test

asavonic commented:
A few things have changed since Feb:
1. `Copyable and Escapable` seem to be implicitly added to type specifiers. We need to add them to input SIL.
2. The workaround with `class F` now causes a crash after bitcast from float to ptr. I think this workaround was incorrect in the first place, because we substitute `F` with unrelated type `Float`.
3. There is a new crash with a cast to `LoadableTypeInfo`. The current patch already handles some of these cases (check `isa<LoadableTypeInfo>`), when a generic type  (they are not loadable) appears where it is not expected. Unfortunately, I was not able to debug this further.

Here is a WIP patchset that partially addresses (1) and (2), and triggers (3):
https://github.com/asavonic/swift/tree/yield-diff-pa-regression

asl commented:
@swift-ci please test

asl commented:
@swift-ci please test linux

asl commented:
@swift-ci please test windows

asl commented:
@swift-ci please test windows

asl commented:
@swift-ci please test macos

asl committed:
Add autodiff execution test for modify accessor differentiation

asl commented:
@swift-ci please test

asl commented:
I rebased & cleaned this PR, also end-to-end autodiff execution test that relies on this functionality was added.

asl commented:
@rjmccall Will you please review this PR? I believe it is ok as a first approximation, without this `partial_apply` of a coroutine yields an assertion in the compiler.

Thanks!

asl commented:
@swift-ci please test macos

asl commented:
@swift-ci please test

asl commented:
@rjmccall ping

asl commented:
@rjmccall ping

asl commented:
Looks like @rjmccall is unavailable for reviews. @hborla @tkremenek @kavon @xedin Do you happen to know who might be able to review this PR?

Thanks!

tkremenek commented:
> Looks like @rjmccall is unavailable for reviews. @hborla @tkremenek @kavon @xedin Do you happen to know who might be able to review this PR?

Looking for another reviewer to help here!

al45tair commented:
Just FWIW, while @mikeash and I are apparently on the reviewers list for this (as code owners — not sure why though, as it doesn't seem to touch anything we're owners of), I think there are probably better people to review changes in IRGen. I don't know if @aschwaighofer would be able to take a look, maybe?

aschwaighofer commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -6258,6 +6258,7 @@ IRGenModule::getAddrOfContinuationPrototype(CanSILFunctionType fnType) {
   llvm::Function *&entry = GlobalFuncs[entity];
   if (entry) return entry;
 
```
I think that adding a GenericContextScope is generally considered a code smell, as it sets some global state to workaround issues with type substitution.

@slavapestov Am I remembering this right?

aschwaighofer replied:
I am taking this comment "back".

The other Signature expansion APIs add a GenericContextScope as well (e.g in `Signature::getUncached`)

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -1987,6 +2237,8 @@ std::optional<StackAddress> irgen::emitFunctionPartialApplication(
   SmallVector<SILType, 4> argValTypes;
   SmallVector<ParameterConvention, 4> argConventions;
 
+  GenericContextScope scope(IGF.IGM, origType->getInvocationGenericSignature());
```
Another instance of `GenericContextScope`.

If this indeed the right solution, can we guard it behind an autodiff type check (only do it if the type is autodiff)?

asl replied:
@aschwaighofer The PR is not autodiff related per se, though it is required for autodiff as autodiff seems to be the only part of the compiler that produces such SIL currently. 

The PR is about codegen support of `partial_apply` for a co-routine (modify / read accessors for example). Check the testcases, they are all plain SIL, no autodiff is involved.

Coroutines are quite special inside compiler in many aspects. One particular thing is that yields should be substituted in the context of caller, and this causes lots of headaches everywhere.

asl replied:
I tend to agree that GenericContextScope might not be the right solution here. This is why I wanted to have @rjmccall's opinion, how it would be the best way to codegen partial_apply for coroutines given all special cases here.

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2019,13 +2275,20 @@ std::optional<StackAddress> irgen::emitFunctionPartialApplication(
                                                  outType->isNoEscape());
 
+    if (!argLoweringTy->isLegalSILType()) {
+      argLoweringTy = argLoweringTy.subst(subs)->getCanonicalType();
```
Ditto: If this is only needed for autodiff can we guard this code?

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2019,13 +2275,20 @@ std::optional<StackAddress> irgen::emitFunctionPartialApplication(
     if (schema.empty() && !param.isFormalIndirect())
       return;
 
+    if (argType.hasTypeParameter())
```
Ditto: If this is only needed for autodiff can we guard this code?

aschwaighofer commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -6698,8 +6698,12 @@ void IRGenSILFunction::visitConvertFunctionInst(swift::ConvertFunctionInst *i) {
     Explosion res;
-    auto sig = IGM.getSignature(fnType);
-    res.add(Builder.CreateBitCast(fn, sig.getType()->getPointerTo()));
+    auto &fnTI = IGM.getTypeInfoForLowered(fnType);
```
Can we guard this so this only happens for autoDiff functions?

asl replied:
There is nothing autodiff specific here.

asl commented:
@aschwaighofer To give more context on this. Consider the following code:
```llvm
sil public_external @generic_captured_param : $@convention(thin) <T> (Int, @inout T) -> Int

sil @partial_apply_generic_capture : $@convention(thin) (Int) -> @callee_owned (Int) -> Int {
entry(%x : $Int):
  %a = alloc_stack $Int
  store %x to %a : $*Int
  %f = function_ref @generic_captured_param : $@convention(thin) <T> (Int, @inout T) -> Int
  %p = partial_apply %f<Int>(%a) : $@convention(thin) <T> (Int, @inout T) -> Int
  dealloc_stack %a : $*Int
  return %p : $@callee_owned (Int) -> Int
}
```

vs
```llvm
sil @generic_captured_param : $@yield_once @convention(thin) <T> (Int, @inout T) -> (@yields Int);

sil @partial_apply_generic_capture : $@convention(thin) (Int) -> (@yield_once @callee_guaranteed (Int) -> (@yields Int)) {
entry(%x : $Int):
  %a = alloc_stack $Int
  store %x to %a : $*Int
  %f = function_ref @generic_captured_param : $@yield_once @convention(thin) <T> (Int, @inout T) -> (@yields Int)
  %p = partial_apply %f<Int>(%a) : $@yield_once @convention(thin) <T> (Int, @inout T) -> (@yields Int)
  dealloc_stack %a : $*Int
  return %p : $@yield_once @callee_owned (Int) -> (@yields Int)
}
```

The callee types for `%f<Int>` in `partial_apply` would be different:
```
(sil_function_type type="@convention(thin) (Int, @inout Int) -> Int"
  (input=struct_type decl="Swift.(file).Int")
  (input=struct_type decl="Swift.(file).Int")
  (result=struct_type decl="Swift.(file).Int")
  (substitution_map null_generic_signature)
  (substitution_map null_generic_signature))
```
vs
```
(sil_function_type type="@yield_once @convention(thin) @substituted <τ_0_0 where τ_0_0 : Copyable, τ_0_0 : Escapable> (Int, @inout τ_0_0) -> @yields Int for <Int>"
  (input=struct_type decl="Swift.(file).Int")
  (input=generic_type_param_type depth=0 index=0)
  (yield=struct_type decl="Swift.(file).Int")
  (substitution_map generic_signature=<τ_0_0 where τ_0_0 : Copyable, τ_0_0 : Escapable>
    (substitution τ_0_0 ->
      (struct_type decl="Swift.(file).Int"))
    (conformance type="τ_0_0"
      (builtin_conformance type="Int" protocol="Copyable"))
    (conformance type="τ_0_0"
      (builtin_conformance type="Int" protocol="Escapable")))
  (substitution_map null_generic_signature))
```

This is exactly the point of https://github.com/swiftlang/swift/commit/07f03bd287b5f5297ef7a30a49622d0ad29243eb#diff-6914244187ae8314024ba1a6c8b21e5bb788774002728ccc36148e8c3b11a28fR3192 (and below)

So, maybe some other approach should be taken when doing `partial_apply` of a coroutine?

asavonic committed:
Add remaining partial_apply.sil tests for partial_apply_coro.sil

asl commented:
@swift-ci please test

asl commented:
@swift-ci please test

asl commented:
@swift-ci please test

asavonic committed:
Do not emit a context placeholder for @error results

A placeholder goes instead of the context in case of error results, so
that thin functions have the same number of arguments as thick
ones. Coroutine lowering adds it as part of getCallEmission (via
setFromCallee), so we don't need to do it in
emitPartialApplicationForwarder.

asavonic committed:
Emit unnamed coroutine continuations functions

asavonic committed:
Add partial_apply_coro tests with resilient types

asl committed:
Unbreak the test

asl commented:
@swift-ci please test

asl commented on lib/IRGen/Linking.cpp:
```diff
@@ -376,8 +376,10 @@ std::string LinkEntity::mangleAsString() const {
   case Kind::CoroutineContinuationPrototype:
-    return mangler.mangleCoroutineContinuationPrototype(
-                                            cast<SILFunctionType>(getType()));
+    // Emit unnamed functions for coroutine continuation
```
The mangler is unable to handle primary archetypes (as shown in one of the testcases). However, mangling is unnecessary here as coroutine prototype name is not ABI-specific, the coroutine lowering pass in LLVM splits and renames the function anyway, we only need the name to be unique.

aschwaighofer replied:
Okay. 

If we want to get the ability back we can add the following code and reinstate the mangling:

```
--- a/lib/IRGen/GenFunc.cpp
+++ b/lib/IRGen/GenFunc.cpp
@@ -1456,8 +1456,11 @@ public:
         Self(nullptr), FirstData(nullptr), SecondData(nullptr) {}
 
   void begin() override {
+    auto unsubstType = substType->getUnsubstitutedType(IGM.getSILModule());
     auto prototype = subIGF.IGM.getOpaquePtr(
-        subIGF.IGM.getAddrOfContinuationPrototype(origType));
+        subIGF.IGM.getAddrOfContinuationPrototype(
+          cast<SILFunctionType>(
+            unsubstType->mapTypeOutOfContext()->getCanonicalType())));
 
     // Use malloc and free as our allocator.
     auto allocFn = subIGF.IGM.getOpaquePtr(subIGF.IGM.getMallocFn());
```

asl commented on test/AutoDiff/validation-test/modify_accessor.swift:
```diff
@@ -0,0 +1,43 @@
+// RUN: %target-run-simple-swift
```
The example here is not directly related, though the autodiff code currently is the only producer of `partial_apply` of coroutines in the codebase (though, nothing prohibits this per SIL spec).

asl commented:
preset=buildbot,tools=RA,stdlib=DA
@swift-ci test with preset macOS

eeckstein reviewed (commented):
Can you add a paragraph about partial-apply co-routines into `docs/SIL.rst` (under the section for `partial_apply`)? Like you wrote in the description of this PR. Please also add that this is (currently) only used for auto-diff.

Do escaping and non-escaping (`[on_stack]`) `partial_apply`s support co-routines? If not, please document this as well (and add a check in the SIL verifier).

asl commented:
> Do escaping and non-escaping (`[on_stack]`) `partial_apply`s support co-routines? If not, please document this as well (and add a check in the SIL verifier).

This is one of testcases (see https://github.com/swiftlang/swift/pull/71653/files#diff-2eac67a6f96297f40450a009472fdbb4cdb09445c4eb3f9a8b69550e19f2b9b3R600 and below)

```llvm
sil @partial_apply_class_on_stack : $@convention(thin) (@owned SwiftClass) -> () {
entry(%a : $SwiftClass):
  %f = function_ref @partially_applyable_to_class : $@yield_once @convention(thin) (@guaranteed SwiftClass) -> (@yields SwiftClass)
  %c = partial_apply [callee_guaranteed] [on_stack] %f(%a) : $@yield_once @convention(thin) (@guaranteed SwiftClass) -> (@yields SwiftClass)
  %use = function_ref @use_closure : $@convention(thin) (@noescape @yield_once @callee_guaranteed () -> (@yields SwiftClass)) -> ()
  apply %use(%c) : $@convention(thin) (@noescape @yield_once @callee_guaranteed () -> (@yields SwiftClass)) -> ()
  dealloc_stack %c : $@noescape @yield_once @callee_guaranteed () -> (@yields SwiftClass)
  strong_release %a : $SwiftClass
  %t = tuple()
  return %t : $()
}
```

So it seems to be there

asl committed:
Fix invalid pointer authentication lowering for coroutine partial apply.
Fix tests on arm64e

asl committed:
Add documentation note for partial_apply of a coroutine

asl commented:
@swift-ci please test

asl commented:
@eeckstein I added the note to docs in https://github.com/swiftlang/swift/pull/71653/commits/962777fba96dc40fd70548a33fb3fce7d135ffc0

Please let me know if this is enough or something else should be added

asl commented:
preset=buildbot,tools=RA,stdlib=RD,test=non_executable
@swift-ci Please test with preset macOS

eeckstein commented:
> I added the note to docs in https://github.com/swiftlang/swift/commit/962777fba96dc40fd70548a33fb3fce7d135ffc0

Perfect, thanks!

asl committed:
Refactor all the tests to be robust wrt platform differences

asl commented:
@swift-ci please test

asl commented:
preset=buildbot,tools=RA,stdlib=RD,test=non_executable
@swift-ci Please test with preset macOS

aschwaighofer reviewed (approved):
LGTM

asl committed:
Bring mangling of coroutine continuations back.

Co-authored-by: Arnold Schwaighofer <aschwaighofer@apple.com>

asl commented:
@swift-ci please test

asl commented:
Buildbot configiration fails due `Parse/enum_floating_point_raw_value.swift` test broken on non-x86 targets after 66104395d74e2b93cad01e95e679b2304b78552a and is unrelated to this PR
